### PR TITLE
Array Backend Part 2: Signed Int implementation

### LINKED ISSUE #92: inconsistency between stint[>=128] and stint[<=64] during creation from native signed int
stint[8] to stint[64] allow creation from low(int8-64), but stint[128] and above will fail with overflow exception.

for example:
```Nim
let a = low(int8).stint(128)
let b = low(int16).stint(128)
let c = low(int32).stint(128)
let d = low(int64).stint(128)
```

### LINKED ISSUE #98: compile time issue with nim v1.0.2
currently there are 2 bugs discovered, but it requires complex/long code to reproduce it. and they also have possibility entangled.

I open this issue as a reminder to fix it in the future.

more information will be added after I understand the problem more precisely.

affected proc: `powmod` of modular_arithmetic.nim
current action: disable compile time for `powmod`

### PR DESCRIPTION
**Pull request by jangko**
fix #92 
fix #98

### CODE REVIEW

File: stint/intops.nim [hunk 1]
```diff
@@ -7,155 +7,355 @@
...
+
+{.pop.}
+
+# Addsub
+# --------------------------------------------------------
+{.push raises: [], inline, noinit, gcsafe.}
+
+func `+`*(a, b: StInt): StInt =
+  ## Addition for multi-precision unsigned int
+  result.imp.sum(a.imp, b.imp)
```
Comment (mratsim) [inline stint/intops.nim:207]: Are there overflow checks?

File: stint/intops.nim [hunk 2]
```diff
@@ -7,155 +7,355 @@
...
+  ## In-place addition for multi-precision unsigned int
+  a.imp.sum(a.imp, b.imp)
 
-func high*[bits](_: typedesc[StInt[bits]]): StInt[bits] {.inline.} =
-  result.data = high(type result.data)
-func high*[bits](_: typedesc[StUint[bits]]): StUint[bits] {.inline.} =
-  result.data = high(type result.data)
+func `-`*(a, b: StInt): StInt =
+  ## Substraction for multi-precision unsigned int
+  result.imp.diff(a.imp, b.imp)
```
Comment (mratsim) [inline stint/intops.nim:215]: underflow checks?

File: stint/intops.nim [hunk 3]
```diff
@@ -7,155 +7,355 @@
...
-  result.data = low(type result.data)
-func low*[bits](_: typedesc[StUint[bits]]): StUint[bits] {.inline.} =
-  result.data = low(type result.data)
+func `-=`*(a: var StInt, b: StInt) =
+  ## In-place substraction for multi-precision unsigned int
+  a.imp.diff(a.imp, b.imp)
 
-import ./private/uint_exp, math
+func inc*(a: var StInt, w: Word = 1) =
+  a.imp.inc(w)
```
Comment (mratsim) [inline stint/intops.nim:222]: overflow checks?

File: stint/intops.nim [hunk 4]
```diff
@@ -7,155 +7,355 @@
...
+    tmp = d.neg
+    result.imp = nneg.imp div tmp.imp
+    return
+
+  # neg / pos
+  result.imp = nneg.imp div d.imp
+  result.negate
+
+func `mod`*(x, y: StInt): StInt =
+  ## Remainder operation for multi-precision unsigned uint
```
Comment (mratsim) [inline stint/intops.nim:302]: The description should be clear of the sign of the remainder due to confusion between C and mathematical modulo/remainder.

Also whatever is chosen we should have `x = q*m+r` with q the quotient and r the remainder/modulo and m some multiple.

File: stint/private/datatypes.nim [hunk 5]
```diff
@@ -40,7 +40,7 @@ type
   StInt*[bits: static[int]] = object
     ## Stack-based integer
     ## Signed
-    limbs*: array[bits.wordsRequired, Word]
+    imp*: StUint[bits]
```
Comment (mratsim) [inline stint/private/datatypes.nim:43]: nitpick: `impl` is probably better than `impl`

**Review by mratsim — APPROVED** [review]
SGTM, great work.

Though if we want a superset of signed integers, the signed integers need overflow and underflow checks.

**Comment by jangko** [comment]
@mratsim thanks for the inputs, overflow check will be added in part 3. currently my focus is to make stint works with nimbus-eth1 and nimbus-eth2.

File: tests/test_int_bitwise.nim [hunk 6]
```diff
@@ -257,29 +100,6 @@ template testBitwise(chk, tst: untyped) =
     chkShl(chk, "0F", 255, "8000000000000000000000000000000000000000000000000000000000000000", 256)
 
   tst "operator `shr`":
-    chkShr(chk, "0F", 4, "00", 8)
```
Comment (arnetheduck) [inline tests/test_int_bitwise.nim:260]: why the removal of these tests?

Comment (jangko) [reply tests/test_int_bitwise.nim]: they are for stuint[8-32], I don't think they are used.

**Pull request merged**